### PR TITLE
Implement BigInteger.GetBitLength

### LINKED ISSUE #31308: Add API to get bit length of BigInteger
Gets the number of bits required for shortest two's complement representation of a `BigInteger` instance without the sign bit.

The method returns the minimum non-negative number of bits in two's complement notation without the sign bit.

This method returns `0` _iff_ value of the `BigInteger` instance is equal to `0` or `-1`. For positive integers the return value is equal to the ordinary binary representation string length.

[Java docs for equivalent method](https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/math/BigInteger.html#bitLength())

Implementation implies zero impact on existing code, maintenance cost is expected to be low.

# Current story
```cs
private static long SlowBitLength(BigInteger integer)
{
    return (long)Math.Ceiling(BigInteger.Log(integer.Sign < 0 ? -integer : integer + 1, 2.0));
}
```

* Slower
* Imprecise for huge (`2^(32*int.MaxValue)`) values (`correct ± 1`)

# Proposed API
```cs
public struct BigInteger : <...>
{
	<...>
    public long GetBitLength();
	<...>
}
```

# Sample results
samples

# Open Questions
* Should it be property, or method? It implies some computation, so it's cheap, not free. `IsPowerOfTwo` is a property and has approximately the same complexity. — **Chosen method approach.**

# References
dotnet/runtime#29193 contains a similar method as one of the points, however this method is different.

# Changelog
* [2019-11-02] Rewritten as method (was property initially).
* [2020-06-15] Fixed description to take two's complement representation as the basis.

### PR DESCRIPTION
**Pull request by andrew-boyarshin**
Closes #31308.

### CODE REVIEW

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs [hunk 1]
```diff
@@ -2303,6 +2303,37 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
...
+                int sign = _sign;
+                highValue = sign < 0 ? (uint)-sign : (uint)sign;
+            }
+            else
+            {
+                byteLength = bits.Length;
+                highValue = bits[byteLength - 1];
+            }
+
+            int highZeros = NumericsHelpers.CbitHighZero(highValue);
```
Comment (tannergooding) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:2332]: It would be good to use `BitOperations.LeadingZeroCount` instead.
Ideally we would switch all usages of `CbitHighZero` to `BitOperations.LeadingZeroCount` but that should likely be a separate PR.

Comment (andrew-boyarshin) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:2332]: This is an optional recommendation, right?
Suggested options:
* edit it (edits by maintainers allowed)
* wait for me to apply the change somewhere in mid-July (unfortunately I've got no free time at the moment and I'm not familiar with `BitOperations` to do it right now without proper testing)
* merge as is, switching it to use `BitOperations.LeadingZeroCount` with the rest of the `CbitHighZero` usages

Comment (tannergooding) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:2332]: Right, it isn't required. If you aren't able to handle this particular scenario I'll log a bug and have it done as part of fixing all `CbitHighZero` usages.

Comment (andrew-boyarshin) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:2332]: Nvm, I've already rearranged my schedule. I'll try to do it right now.

Comment (andrew-boyarshin) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:2332]: Turns out it's a direct replacement.

**Comment by andrew-boyarshin** [comment]
Congratulations, .NET `BigInteger` is now bigger than Java one. 😄
`BigInteger.bitLength` in Java returns an `int`, whereas in .NET 'BigInteger.GetBitLength' will return `long`, and for a good reason: thanks to the issue first raised as #15925, it has been discovered that, while theoretical and rarely seen in practice, .NET `BigInteger` is capable of handling e.g. 2^(64*`int.MaxValue`). In Java BCL that's not the case, see [test in JDK repo](https://github.com/openjdk/jdk15/blob/master/test/jdk/java/math/BigInteger/BitLengthOverflow.java) (if the test passes and gives a correct result, that's a surprise ~_for sure, but a welcome one_~).

P.S. I've also discovered that `SlowBitLength` gives an incorrect answer (`correct - 1`) on such large cases. I don't know if this can be fixed (I think that's floating point precision playing its tricks).

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs [hunk 2]
```diff
@@ -2303,6 +2303,35 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
...
+                int sign = _sign;
+                highValue = sign < 0 ? (uint)-sign : (uint)sign;
+            }
+            else
+            {
+                byteLength = bits.Length;
+                highValue = bits[byteLength - 1];
+            }
+
+            return byteLength * 32L - BitOperations.LeadingZeroCount(highValue);
```
Comment (KalleOlaviNiemitalo) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:2332]: The `byteLength` name seems misleading because it counts 32-bit units and `System.Byte` is 8-bit.

File: src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs [hunk 3]
```diff
@@ -2303,6 +2303,35 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
             return !right.Equals(left);
         }
 
+        /// <summary>
+        /// Gets the number of bits required to represent this BigInteger instance in two's complement notation without the sign bit.
+        /// </summary>
+        /// <returns>The non-negative number of bits in two's complement notation without the sign bit.</returns>
```
Comment (KalleOlaviNiemitalo) [inline src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs:2309]: In two's complement notation, -2 is …1110<sub>2</sub>; all the ones are sign bits, so I'd expect the method to return 1. From the tests however, it appears that `x.GetBitLength() == (-x).GetBitLength()` always. I am not sure that "two's complement notation" correctly describes the behavior of this method. If the documentation instead talked about the number of bits in the absolute value, then it would not need to say anything about sign bits.

Comment (andrew-boyarshin) [reply src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: Thanks! The two's complement representation is actually the intended detail, missing from the implementation (and quite honestly, from the API review).

**Comment by andrew-boyarshin** [comment]
NOMERGE until API proposal re-review.

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
CC. @pgovind, could you give this a second set of eyes?

**Review by pgovind — APPROVED** [review]
LGTM!
